src/HallPIC.jl

```diff
@@ -455,4 +457,107 @@ function boltzmann_electric_field_and_potential!(E, phi, n_e, T_e, grid)
+    coefficient::Int8 
+end
+struct Reaction{T<:AbstractFloat}
+    reactant::ReactingSpecies
+    products::Vector{ReactingSpecies}
+    threshold_energy::T
+    energies::Vector{T}
+    rate::Vector{T}
+    delta_n::Vector{T}
+
+end
+
+
+
+function reaction_reduction(grid::Grid, reaction::Reaction{T}, electron_properties::SpeciesProperties{T}, fluid_properties::SpeciesProperties{T}, reactant::ParticleContainer{T},  dt) where T
+
+    #for each (non-ghost) cell, calculate the delta n
+    for i in 2:length(grid.cell_centers)-1
+        #calculate the number of particles produced
+        rate_idx = findfirst(electron_properties.temp[i] .<= reaction.energies)
+        rate = (reaction.rate[rate_idx] - reaction.rate[rate_idx-1]) / (reaction.energies[rate_idx] - reaction.energies[rate_idx-1]) * (electron_properties.temp[i] - reaction.energies[rate_idx-1]) + reaction.rate[rate_idx-1]
+        #to maintain the normalization, the two densities need to be multiplied by n_0 and the entire addition divided by n_0, results in a net multiplication of n_0
+        reaction.delta_n[i] = fluid_properties.dens[i] * electron_properties.dens[i] * rate * dt * n_0
+    end
+
+    #now that we have the delta_n, adjust particle weights 
+    for i in 2:length(reactant.pos)-1
+        #pull the index 
+        cell_index = reactant.inds[i]
+
+        # don't deposit into ghost cells
+        if cell_index == -2 || cell_index == length(grid.cell_centers)-1
+            s, ic = 0, abs(cell_index)
+        else
+            s, ic = sign(cell_index), abs(cell_index)
+        end
+
+        #count the first cell 
+        reactant.weight[i] -= reactant.weight[i] * (reaction.reactant.coefficient * reaction.delta_n[ic]/fluid_properties.dens[ic]) * abs(reactant.pos[i] - grid.cell_centers[ic]) / grid.dz
+        #count the second cell 
+        reactant.weight[i] -= reactant.weight[i] * (reaction.reactant.coefficient * reaction.delta_n[ic+s]/fluid_properties.dens[ic+s]) * abs(reactant.pos[i] - grid.cell_centers[ic+s]) / grid.dz
+    end
+
+    return reaction, reactant 
+end
+
+"""
+Note that the products vector should be sorted the same way as in the reaction structure 
+"""
+
+function daughter_particle_generation(grid::Grid, reaction::Reaction{T}, reactant_properties::SpeciesProperties{T}, product_properties::SpeciesProperties{T}, products::Vector{ParticleContainer{T}}) where T
+
+
+    dz = grid.face_centers[2:end] - grid.face_centers[1:end-1]
+    #for each product 
+    for p in 1:length(products) 
+        n_d = products[p].n_d#number of desired particles touching cell, hyperparamter from the simualtion
+        #for each cell, add particles 
+        for i=2:length(grid.cell_centers)-1
+            #determine the number of partices to generate
+            w_gen = product_properties.avg_weight[i] * ((sum(products[p].inds.==i)+sum(products[p].inds.==-i)+sum(products[p].inds.==i-1)+sum(products[p].inds .== -(i+1)))/n_d)#check for particles in the cell 
```

Comment:
numpy indexing - avoid. do a loop or a loop comprehension. this is also accidentally quadratic-time (you're iterating over cell centers and in each cell you iterate over every single particle in the domain (four times)). try and think about how you might rearrange this to avoid this. An explicit loop will make it clearer to you how the data is moving and how you might fix the algorithmic complexity.

Reply:
You should be able to get this quantity from the field quantities